lib/initialize_with_config.rb

```diff
@@ -64,17 +66,31 @@ def call(*)
   end
 
   def default_config
-    HashWithIndifferentAccess.new
+    {}
+  end
+
+  def required_config
+    self.class.required_config
   end
 
 private
 
   # @param [Hash] config
   def initialize(config)
-    self.config = config.respond_to?(:[]) ? default_config.merge(config) : default_config
+    self.class.externally_injected ||= []
```

Comment:
InitializeWithConfig is a Dependency injector. externally_injected is a private array which holds items (methods, classes, variables etc.), which have been explicitly declared as required to be present in the config. The main use of this is to compare what has been passed in with what is required and to raise an appropriate error message if anything is missing. They are held against the including class themselves, since that's where they are defined.

Reply:
Right. I've read through a bunch of code, and what I have in my head is:

When you have `SomeClass` include `InitializeWithConfig`, then...

At the moment you make a new instance of `SomeClass`, you add a method to the new instance for each of the config elements. And to `SomeClass` instance, those methods are `externally_injected`?

Reply:
The ones that you have declared in the class with `required_config` are the ones that are compared against the `externally_injected` array. It's mainly to clean up the error messages since it's more meaningful to raise: `::InitializeWithConfig::MissingRequiredArguments` and show which values are missing on instantiation and for which class than to let the Ruby standard ArgumentMissing raise when they are used.